tokio-util/src/sync/poll_semaphore.rs

```diff
@@ -55,10 +53,20 @@ impl PollSemaphore {
     /// the `Waker` from the `Context` passed to the most recent call is
     /// scheduled to receive a wakeup.
     pub fn poll_acquire(&mut self, cx: &mut Context<'_>) -> Poll<Option<OwnedSemaphorePermit>> {
-        let result = ready!(self.permit_fut.poll(cx));
+        let permit_future = loop {
+            match self.permit_fut.as_mut() {
+                Some(fut) => break fut,
+                None => {
+                    let next_fut = Arc::clone(&self.semaphore).acquire_owned();
+                    self.permit_fut = Some(ReusableBoxFuture::new(next_fut));
+                }
```

Comment:
We could even make the `try_acquire_owned` call here, avoiding the allocation when a permit is immediately available.